### PR TITLE
Add 'ispyb.job' to create/edit/view processing jobs

### PR DESCRIPTION
**Pull request by Anthchirp**
A new command, copied over from dlstbx with a few modifications

### CODE REVIEW

File: src/ispyb/cli/job.py [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+
+def run():
+    parser = OptionParser(
+        usage="ispyb.job [options] rpid",
+        description="Command line tool to manipulate ISPyB processing table entries.",
+    )
+
+    available_recipes = filter(
+        lambda r: r.startswith("ispyb-") and r.endswith(".json"),
+        os.listdir("/dls_sw/apps/zocalo/live/recipes"),
```
Comment (stufisher) [inline src/ispyb/cli/job.py]: this is catered for by zocalo.configuration, but thats still to be reviewed

c.f.
https://github.com/DiamondLightSource/python-zocalo-ispyb/blob/main/zocalo_ispyb/cli/ispyb_job.py#L138

https://github.com/DiamondLightSource/python-zocalo/pull/52/files#diff-a00122aee2a4c3b731fcc1d940db7a3737c99d09611afd3d909a5a453af711ebR3

Comment (Anthchirp) [reply src/ispyb/cli/job.py]: Made this code block optional :+1:

**Comment by Anthchirp** [comment]
Reasonably happy with it. But it is untested right now. So needs a test and some documentation maybe

**Comment by Anthchirp** [comment]
- [x] This PR will need a tiny adjustment when #138 is merged

File: src/ispyb/cli/job.py [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+        usage="ispyb.job [options] JOBID",
+        description="Command line tool to manipulate ISPyB processing table entries.",
+    )
+
+    if os.path.isdir("/dls_sw/apps/zocalo/live/recipes"):
+        available_recipes = filter(
+            lambda r: r.startswith("ispyb-") and r.endswith(".json"),
+            os.listdir("/dls_sw/apps/zocalo/live/recipes"),
+        )
+        available_recipes = sorted(r[6:-5] for r in available_recipes)
```
Comment (rjgildea) [inline src/ispyb/cli/job.py]: Could the location of this file be provided via a command line parameter (or alternatively an environment variable)? This _could_ default to the `/dls_sw/` location, but I'm open to arguments against this. See e.g. the `--credentials` parameter to `ispyb.last_data_collections_on` (which will now default to reading from the `ISPYB_CREDENTIALS` environment variable if not explicitly provided on the command line):
https://github.com/DiamondLightSource/ispyb-api/blob/7e9f504f6765a7070a4680e4c2a286510f5da6db/src/ispyb/cli/last_data_collections_on.py#L128

File: src/ispyb/cli/job.py [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+        jisp["datacollectionid"] = sweep[0]
+        jisp["start_image"] = sweep[1]
+        jisp["end_image"] = sweep[2]
+        jispid = i.mx_processing.upsert_job_image_sweep(list(jisp.values()))
+        print(f"  JISP={jispid}")
+
+    print(f"All done. Processing job {jobid} created")
+    print()
+    if zocalo:
+        if options.trigger:
```
Comment (rjgildea) [inline src/ispyb/cli/job.py:114]: Could this be `
```suggestion
    if zocalo and options.trigger:
```
and remove one level of indentation? It seems no harm in printing out the `zocalo.go` command even if zocalo is not available.

We're also using the presence of the `zocalo` module in the current python environment as a proxy for the presence of the `zocalo.go` command line program - is it possible that someone could have zocalo available in a _different_ Python environment?

File: src/ispyb/cli/job.py [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+#   ispyb.job 73 -v  # show full record
+
+# Create new processing program row:
+#   ispyb.job 73 -c -p "program" -s "starting up..."
+
+# Update stored information:
+#   ispyb.job 73 -u 1234 -s "running..."
+#   ispyb.job 73 -u 1234 -s "things are happening" --update-time "2017-08-25"
+#   ispyb.job 73 -u 1234 -s "completed successfully" -r success
+#   ispyb.job 73 -u 1234 -s "everything is broken" -r failure
```
Comment (rjgildea) [inline src/ispyb/cli/job.py]: This should probably be a module-level docstring and/or command line help string.

File: src/ispyb/cli/job.py [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+    jp["datacollectionid"] = dcid or sweeps[0][0]
+    jp["display_name"] = options.display
+    jp["recipe"] = options.recipe
+    print("Creating database entries...")
+
+    jobid = i.mx_processing.upsert_job(list(jp.values()))
+    print(f"  JobID={jobid}")
+    for key, value in parameters:
+        jpp = i.mx_processing.get_job_parameter_params()
+        # _job_parameter_params = StrictOrderedDict([('id', None), ('job_id', None), ('parameter_key', None), ('parameter_value', None)])
```
Comment (rjgildea) [inline src/ispyb/cli/job.py]: Probably not worth carrying over these comments:
```suggestion
```

File: src/ispyb/cli/job.py [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+        jpp["job_id"] = jobid
+        jpp["parameter_key"] = key
+        jpp["parameter_value"] = value
+        jppid = i.mx_processing.upsert_job_parameter(list(jpp.values()))
+        print(f"  JPP={jppid}")
+
+    for sweep in sweeps:
+        jisp = i.mx_processing.get_job_image_sweep_params()
+        # _job_image_sweep_params = StrictOrderedDict([('id', None), ('job_id', None), ('datacollectionid', None), ('start_image', None), ('end_image', None)])
```
Comment (rjgildea) [inline src/ispyb/cli/job.py]: As above
```suggestion
```

File: src/ispyb/cli/job.py (hunk 3 above)
Comment (rjgildea) [reply src/ispyb/cli/job.py:114]: I see now that the `--trigger` option is only set if `zocalo` is available, so perhaps the nested indentation does make sense after all...

**Review by rjgildea — CHANGES_REQUESTED** [review]
It's probably a good idea to move this command here as it only has minimal dependency on zocalo. As @Anthchirp points out we should probably use the opportunity to add tests and docstrings while we're at it.

File: src/ispyb/cli/job.py (hunk 3 above)
Comment (Anthchirp) [reply src/ispyb/cli/job.py:114]: there is also an `else` clause for the inner `if`

File: src/ispyb/cli/job.py (hunk 2 above)
Comment (Anthchirp) [reply src/ispyb/cli/job.py]: Ideally this sort of stuff would be resolved using a zocalo configuration yaml file that not only sets a recipe location, but also stuff like graylog logging, access to activemq/rabbitmq credentials, etc.

I suppose temporary gorilla variables are fine, but I would only like to add them only if someone other than us is _actually_ using them

Comment (Anthchirp) [reply src/ispyb/cli/job.py]: I did however take the opportunity to clear up this idiosyncratic filter lambda generator :slightly_smiling_face:

File: src/ispyb/cli/job.py (hunk 4 above)
Comment (Anthchirp) [reply src/ispyb/cli/job.py]: done

**Comment by Anthchirp** [comment]
No idea if this is valid RST, but it looks plausible to me.

**Review by rjgildea — APPROVED** [review]
It would be nice to add a test for creating/updating entries, but this PR already gives us more test coverage than we currently have, so I guess it's not critical.

**Pull request merged**